src/main/java/com/limechain/network/protocol/warp/dto/WarpSyncJustification.java

```diff
@@ -24,4 +41,94 @@ public String toString() {
                 '}';
     }
+
+    public boolean verify(Authority[] authorities, BigInteger authoritiesSetId) {
+        // Implementation from: https://github.com/smol-dot/smoldot
+        // lib/src/finality/justification/verify.rs
+        if (precommits.length < (authorities.length * 2 / 3) + 1) {
+            log.log(Level.WARNING, "Not enough signatures");
+            return false;
+        }
+
+        Set<Hash256> seenPublicKeys = new HashSet<>();
+        Set<Hash256> authorityKeys = Arrays.stream(authorities)
+                .map(Authority::getPublicKey)
+                .collect(Collectors.toSet());
+
+        for (Precommit precommit : precommits) {
+            if (!authorityKeys.contains(precommit.getAuthorityPublicKey())) {
+                log.log(Level.WARNING, "Invalid Authority for precommit");
+                return false;
+            }
+
+            if (seenPublicKeys.contains(precommit.getAuthorityPublicKey())) {
+                log.log(Level.WARNING, "Duplicated signature");
+                return false;
+            }
+            seenPublicKeys.add(precommit.getAuthorityPublicKey());
+
+            // TODO (from smoldot): must check signed block ancestry using `votes_ancestries`
+
+            // 1 reserved byte for data type
+            // 32 reserved for target hash
+            // 4 reserved for block number
+            // 8 reserved for justification round
+            // 8 reserved for set id
+            int messageCapacity = 1 + 32 + 4 + 8 + 8;
+            var messageBuffer = ByteBuffer.allocate(messageCapacity);
+            messageBuffer.order(ByteOrder.LITTLE_ENDIAN);
+
+            // Write message type
+            messageBuffer.put((byte) 1);
+            // Write target hash
+            messageBuffer.put(LittleEndianUtils
+                    .convertBytes(StringUtils.hexToBytes(precommit.getTargetHash().toString())));
+            //Write Justification round bytes as u64
+            messageBuffer.put(LittleEndianUtils
+                    .bytesToFixedLength(precommit.getTargetNumber().toByteArray(), 4));
+            //Write Justification round bytes as u64
+            messageBuffer.put(LittleEndianUtils.bytesToFixedLength(this.round.toByteArray(), 8));
+            //Write Set Id bytes as u64
+            messageBuffer.put(LittleEndianUtils.bytesToFixedLength(authoritiesSetId.toByteArray(), 8));
+
+            //Verify message
+            //Might have problems because we use the stand ED25519 instead of ED25519_zebra
+            messageBuffer.rewind();
+            byte[] data = new byte[messageBuffer.remaining()];
+            messageBuffer.get(data);
+            boolean isValid = verifySignature(precommit.getAuthorityPublicKey().toString(),
+                    precommit.getSignature().toString(), data);
+            if (!isValid) {
+                log.log(Level.WARNING, "Failed to verify signature");
+                return false;
+            }
```

Comment:
This function takes way too much space, it would be better if we can split it by putting the highlighted part ± in a separate private function.